### DISCUSSION TITLE
P132 (*http.Server).Shutdown メソッドにわたす context.Context について

### DISCUSSION THREAD

**Discussion by ghost** [budougumi0617/go_todo_app#86] [Q&A] [answered]
はじめまして。
「詳解Go言語Webアプリケーション開発」を購入し、読ませていただきました。情報がぎっしり詰まっていて、とても勉強になりました。

本の内容について一つお聞きしたいことがございます。
[ここ](https://github.com/budougumi0617/go_todo_app/blob/main/server.go#L43)のコードの `(*http.Server).Shutdown` メソッドの引数ですが、 `Run` 関数の引数として受け取った `ctx` ではなく、 `context.Background()` で作成した新しいコンテキストを引数として渡しているのはなぜでしょうか？

よろしくお願いします。

**Comment by budougumi0617** [comment, marked as answer]
@ynm3n お読みいただきありがとうございました！なにかお役に立てていれば幸いです！

質問についてですが、 `Run` メソッドで受け取った `ctx` はL42の `<-ctx.Done()` を通過している時点で終了済みのコンテキストになっています。
終了済みのコンテキストをメソッドや関数にわたすと期待する動作をする前に処理が終了してしまうことがあるため、新しいコンテキストを渡しています。
https://github.com/budougumi0617/go_todo_app/blob/v1.0.13/server.go#L42-L47
```go
	<-ctx.Done()
	if err := s.srv.Shutdown(context.Background()); err != nil {
		log.Printf("failed to shutdown: %+v", err)
	}
```

**Reply by ghost** [reply]
`Shutdown` メソッドにそれまでのコンテキストとは別のコンテキストを渡さなければならない理由が理解できました。ご回答ありがとうございました！

**Reply by budougumi0617** [reply]
またご不明点あればお気軽にご質問ください！！
